File: app/src/main/java/bogomolov/aa/fitrack/view/TagSelectionDialog.java
package bogomolov.aa.fitrack.view;

import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.ActionMode;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.Toolbar;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.DialogFragment;
import androidx.lifecycle.ViewModelProviders;

import javax.inject.Inject;

import bogomolov.aa.fitrack.R;
import bogomolov.aa.fitrack.dagger.ViewModelFactory;
import bogomolov.aa.fitrack.databinding.FragmentTagSelectionBinding;
import bogomolov.aa.fitrack.model.Tag;
import bogomolov.aa.fitrack.viewmodels.TagSelectionViewModel;
import dagger.android.support.AndroidSupportInjection;

public class TagSelectionDialog extends DialogFragment {
    private ActionMode actionMode;
    private Toolbar toolbar;
    private ListView listView;

    private Tag selectedTag;
    private TagResultListener tagResultListener;

    private TagSelectionViewModel viewModel;

    @Inject
    ViewModelFactory viewModelFactory;

    @Override
    public void onAttach(@NonNull Context context) {
        AndroidSupportInjection.inject(this);
        super.onAttach(context);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(TagSelectionViewModel.class);
        FragmentTagSelectionBinding viewBinding = DataBindingUtil.inflate(inflater,R.layout.fragment_tag_selection,container,false);
        viewBinding.setViewModel(viewModel);
        viewBinding.setLifecycleOwner(this);
        View view = viewBinding.getRoot();

        listView = view.findViewById(R.id.tag_list_view);
        ArrayAdapter<Tag> adapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_list_item_1, viewModel.getTags());
        listView.setAdapter(adapter);
        viewModel.tagsLiveData.observe(this, (tags) -> adapter.notifyDataSetChanged());

        toolbar = view.findViewById(R.id.tag_selection_toolbar);
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                dismiss();
            }
        });

        listView.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
            @Override
            public boolean onItemLongClick(AdapterView<?> adapterView, View view, int i, long l) {
                if (actionMode == null) {
                    actionMode = toolbar.startActionMode(callback);
                } else {
                    actionMode.finish();
                }
                return true;
            }
        });

        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int position, long l) {
                selectedTag = (Tag) listView.getItemAtPosition(position);
                dismiss();
            }
        });

        return view;
    }

    public void setTagResultListener(TagResultListener tagResultListener) {
        this.tagResultListener = tagResultListener;
    }


    private ActionMode.Callback callback = new ActionMode.Callback() {

        public boolean onCreateActionMode(ActionMode mode, Menu menu) {
            mode.getMenuInflater().inflate(R.menu.tag_selection_menu, menu);
            return true;
        }

        public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
            return false;
        }

        public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
            Tag tag = (Tag) listView.getSelectedItem();
            viewModel.deleteTag(tag);
            return true;
        }

        public void onDestroyActionMode(ActionMode mode) {
            actionMode = null;
        }

    };


    public void onDismiss(DialogInterface dialog) {
        super.onDismiss(dialog);
        if (tagResultListener != null) tagResultListener.onTagSelectionResult(selectedTag);
    }

    public void onCancel(DialogInterface dialog) {
        super.onCancel(dialog);
        if (tagResultListener != null) tagResultListener.onTagSelectionResult(null);
    }

}

File: app/src/main/java/bogomolov/aa/fitrack/features/tracks/tags/TagSelectionDialog.kt
package bogomolov.aa.fitrack.features.tracks.tags

import android.content.DialogInterface
import android.os.Bundle
import android.view.*
import android.widget.ArrayAdapter
import android.widget.ListView
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.DialogFragment
import bogomolov.aa.fitrack.R
import bogomolov.aa.fitrack.databinding.FragmentTagSelectionBinding
import bogomolov.aa.fitrack.domain.model.Tag
import org.koin.androidx.viewmodel.ext.android.viewModel

class TagSelectionDialog(private val onTagSelection: (Tag?)->Unit) : DialogFragment() {
    private val viewModel: TagSelectionViewModel by viewModel()
    private var actionMode: ActionMode? = null
    private lateinit var toolbar: Toolbar
    private lateinit var listView: ListView
    private var selectedTag: Tag? = null
    private var selectedToDeleteTag: Tag? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val binding = FragmentTagSelectionBinding.inflate(inflater, container, false)

        listView = binding.tagListView
        viewModel.tagsLiveData.observe(this) { tags ->
            val adapter = ArrayAdapter(
                requireContext(),
                android.R.layout.simple_list_item_1,
                tags.map { it.name })
            listView.adapter = adapter
        }

        toolbar = binding.tagSelectionToolbar
        toolbar.setTitle(R.string.title_select_tag)

        val closeButton = binding.tagsBackButton
        closeButton.setOnClickListener { dismiss() }

        listView.setOnItemLongClickListener { _, _, position, _ ->
            if (actionMode == null) {
                actionMode = toolbar.startActionMode(callback)
                selectedToDeleteTag = viewModel.tagsLiveData.value!![position]
            } else {
                actionMode!!.finish()
            }
            true
        }

        listView.setOnItemClickListener { _, _, position, _ ->
            if (actionMode == null) {
                selectedTag = viewModel.tagsLiveData.value?.get(position)
                dismiss()
            } else {
                selectedToDeleteTag = viewModel.tagsLiveData.value?.get(position)
            }
        }

        binding.tagNameEditLayout.setEndIconOnClickListener {
            val tagName = binding.tagNameEditText.text.toString()
            viewModel.onNewTag(tagName)
            binding.tagNameEditText.setText("")
        }

        return binding.root
    }

    override fun onDismiss(dialog: DialogInterface) {
        super.onDismiss(dialog)
        onTagSelection(selectedTag)
    }

    override fun onCancel(dialog: DialogInterface) {
        super.onCancel(dialog)
        onTagSelection(null)
    }

    private val callback = object : ActionMode.Callback {

        override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
            mode.menuInflater.inflate(R.menu.tag_selection_menu, menu)
            return true
        }

        override fun onPrepareActionMode(mode: ActionMode, menu: Menu) = false

        override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
            if (selectedToDeleteTag != null) viewModel.deleteTag(selectedToDeleteTag!!)
            actionMode?.finish()
            return true
        }

        override fun onDestroyActionMode(mode: ActionMode) {
            actionMode = null
        }
    }
}